### PR TITLE
feat: Add a limit to the number of bytes that can be outstanding to a publisher

### PR DESCRIPTION
**Pull request by palmere-google**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-pubsublite-flink/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

### CODE REVIEW

File: src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java [hunk 1]
```diff
@@ -37,6 +40,9 @@
   // Required. The path of the topic to publish messages to.
   public abstract TopicPath topicPath();
 
+  // Optional. The maximum number of bytes a publisher may have outstanding.
```
Comment (dpcollins-google) [inline src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java:43]: nit, don't know if per-sink is implied or should be explicit?

File: src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java [hunk 2]
```diff
@@ -23,10 +23,13 @@
 
 @AutoValue
 public abstract class PubsubLiteSinkSettings<InputT> implements Serializable {
+  public static final int DEFAULT_MAX_BYTES_OUTSTANDING = 10 * 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (dpcollins-google) [reply src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java:26]: also: long

File: src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java [hunk 3]
```diff
@@ -37,6 +40,9 @@
   // Required. The path of the topic to publish messages to.
   public abstract TopicPath topicPath();
 
+  // Optional. The maximum number of bytes a publisher may have outstanding.
+  public abstract Integer maxBytesOutstanding();
```
Comment (dpcollins-google) [inline src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java:44]: long

File: src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java [hunk 4]
```diff
@@ -49,6 +55,9 @@ PublisherOptions getPublisherConfig() {
     // Required.
     public abstract Builder<InputT> setTopicPath(TopicPath value);
 
+    // Optional.
+    public abstract Builder<InputT> setMaxBytesOutstanding(Integer value);
```
Comment (dpcollins-google) [inline src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java:59]: long

File: src/main/java/com/google/cloud/pubsublite/flink/internal/sink/MessagePublisher.java [hunk 5]
```diff
@@ -22,21 +22,40 @@
...
+  private final Semaphore bytesOutstanding;
+
+  public MessagePublisher(Publisher<MessageMetadata> publisher, int maxBytesOutstanding) {
     this.publisher = publisher;
     this.publishes = new ArrayList<>();
+    this.maxBytesOutstanding = maxBytesOutstanding;
+    this.bytesOutstanding = new Semaphore(maxBytesOutstanding);
+  }
+
+  private int getAccountedSize(Message message) {
```
Comment (dpcollins-google) [inline src/main/java/com/google/cloud/pubsublite/flink/internal/sink/MessagePublisher.java:44]: for all of these: just use long instead of int

Comment (dpcollins-google) [reply src/main/java/com/google/cloud/pubsublite/flink/internal/sink/MessagePublisher.java:44]: Ugh semaphore has int permits. never mind.

File: src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java (hunk 4 above)
Comment (dpcollins-google) [reply src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java:59]: Ignore. int instead

File: src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dpcollins-google) [reply src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java:26]: int instead

File: src/main/java/com/google/cloud/pubsublite/flink/internal/sink/MessagePublisher.java [hunk 6]
```diff
@@ -22,21 +22,40 @@
...
+    }
+    return Math.toIntExact(size);
   }
 
   @Override
-  public void publish(Message message) {
-    publishes.add(publisher.publish(message));
+  public void publish(Message message) throws InterruptedException {
+    final int size = getAccountedSize(message);
+    bytesOutstanding.acquire(size);
```
Comment (dpcollins-google) [inline src/main/java/com/google/cloud/pubsublite/flink/internal/sink/MessagePublisher.java:55]: consider adding a tryAcquire first, that logs warning if flow controlled before unconditionally calling acquire?

**Review by dpcollins-google — APPROVED** [review]

File: src/main/java/com/google/cloud/pubsublite/flink/internal/sink/MessagePublisher.java (hunk 6 above)
Comment (palmere-google) [reply src/main/java/com/google/cloud/pubsublite/flink/internal/sink/MessagePublisher.java:55]: Oh, yeah that's a good idea. Done

File: src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (palmere-google) [reply src/main/java/com/google/cloud/pubsublite/flink/PubsubLiteSinkSettings.java:26]: Done

**Pull request merged**